assets/dev/js/frontend/handlers/container/grid-outline.js

```diff
@@ -0,0 +1,224 @@
+
+		this.elements.outlineParentContainer = container.querySelector( boxedContainer ) || container;
+	}
+
+	removeExistingOverlay() {
+		this.elements.gridOutline?.remove();
+	}
+
+	createOverlayContainer() {
+		const { outlineParentContainer } = this.elements,
+			{ classes: { outline } } = this.getDefaultSettings(),
+			gridOutline = document.createElement( 'div' );
+
+		gridOutline.classList.add( outline );
+		outlineParentContainer.appendChild( gridOutline );
+
+		this.elements.gridOutline = gridOutline;
+
+		this.setGridOutlineDimensions();
+	}
+
+	createOverlayItems() {
+		const { gridOutline } = this.elements,
+			{ classes: { outlineItem } } = this.getDefaultSettings(),
+			{ rows, columns } = this.getDeviceGridDimensions(),
+			numberOfItems = rows.length * columns.length;
+
+		for ( let i = 0; i < numberOfItems; i++ ) {
+			const gridOutlineItem = document.createElement( 'div' );
+
+			gridOutlineItem.classList.add( outlineItem );
+			gridOutline.appendChild( gridOutlineItem );
+		}
+	}
+
+	/**
+	 * Get the grid dimensions for the current device.
+	 *
+	 * @return { { columns: { value, length }, rows: { value, length } } }
+	 */
+	getDeviceGridDimensions() {
+		const currentDevice = elementor.channels.deviceMode.request( 'currentMode' );
+
+		return {
+			rows: this.getControlValues( 'grid_rows_grid', currentDevice, 'grid-template-rows' ) || 1,
+			columns: this.getControlValues( 'grid_columns_grid', currentDevice, 'grid-template-columns' ) || 1,
+		};
+	}
+
+	setGridOutlineDimensions() {
+		const { gridOutline } = this.elements,
+			{ rows, columns } = this.getDeviceGridDimensions();
+
+		gridOutline.style.gridTemplateColumns = columns.value;
+		gridOutline.style.gridTemplateRows = rows.value;
+	}
+
+	/**
+	 * Set the control value for the current device.
+	 * Distinguished between grid custom values and slider controls.
+	 *
+	 * @param {string} control - The control name.
+	 * @param {string} device - The device mode.
+	 * @param {string} property - The CSS property name we need to copy from the parent container.
+	 *
+	 * @return {Object} - E,g. {value: repeat(2, 1fr), length: 2}.
+	 */
+	getControlValues( control, device, property ) {
+		const elementSettings = this.getElementSettings(),
+			{ unit, size } = elementSettings[ control ],
+			{ outlineParentContainer } = this.elements,
+			controlValueForCurrentDevice = elementorFrontend.utils.controls.getResponsiveControlValue( elementSettings, control, 'size', device ),
+			controlValue = this.getComputedStyle( outlineParentContainer, property ),
+			computedStyleLength = controlValue.split( ' ' ).length;
+
+		let controlData;
+
+		if ( ( 'custom' === unit && 'string' === typeof controlValueForCurrentDevice ) || size < computedStyleLength ) {
+			controlData = { value: controlValue };
+		} else {
+			// In this case the data taken from the getComputedStyle and not from the control, in order to handle cases that the user have more elements than grid cells.
```

Comment:
Maybe: 'the data is taken'.
Maybe: 'in order to handle cases when the user has more elements.